provider/inmemory.go

```diff
@@ -126,38 +131,50 @@ func (im *InMemoryProvider) validateChangeBatch(zone string, changes *plan.Chang
 	if !ok {
 		return ErrZoneNotFound
 	}
-	mesh := map[string]bool{}
+	mesh := map[string]map[string]bool{}
```

Comment:
totally agree, code looks messy. This is a inmemory dns provider which implements the logic usually sitting on dns provider side, mostly for ease of testing or running external-dns without any real changes but with the ability to track down potential issues, e.g. when two ingress request same hostname. I would probably refactor this code in a separate PR